### PR TITLE
DATA-3345 Update go SDK with exportTabularData

### PR DESCRIPTION
**Pull request by vijayvuyyuru**
Hey yall, this PR updates the Go SDK to include the ExportTabularData endpoint.

### CODE REVIEW

File: app/data_client.go
Comment (purplenicole730) [inline app/data_client.go:1220]: For consistency, this should be `exportTabularDataReturnFromProto`

File: app/data_client.go
Comment (purplenicole730) [inline app/data_client.go:123]: We're trying to stay away from proto types, so `TimeCaptured` should be `time.Time`. There are examples on how the conversion happens in other methods.

File: app/data_client.go
Comment (purplenicole730) [inline app/data_client.go:124]: Similarly, `*anypb.Any` should use an `Any` type that's defined in the app package. I actually have a PR up that's in review right now that introduces this `Any` type, so you can use this type as well.

File: app/data_client.go
Comment (purplenicole730) [inline app/data_client.go:126]: Likewise, `Payload` should be of type `map[string]interface{}`. There are other functions that also convert to this type.

File: app/data_client.go
Comment (purplenicole730) [inline app/data_client.go:1234]: These should be changed to non-proto types. See above comments

**Review by purplenicole730 — CHANGES_REQUESTED** [review]

File: app/data_client.go
Comment (vijayvuyyuru) [reply app/data_client.go:123]: Ah ok, will update. Thanks!

File: app/data_client_test.go
Comment (vijayvuyyuru) [inline app/data_client_test.go:402]: @purplenicole730 One thing that feels a little weird to me is that I have to use the proto response to satisfy all the interfaces for the inject and as a result I'm expecting a proto message instead of the nice golang struct. Do you have any thoughts on that/is what I'm sayin gmaking sense.

Comment (purplenicole730) [reply app/data_client_test.go:402]: I'm not really sure what you mean. Our aim to avoid proto types is to avoid giving the users proto types. If the inject uses the proto type, that's fine since the inject function is not what the users will be using.

File: app/data_client.go
Comment (purplenicole730) [inline app/data_client.go:138]: This can return a pointer to a `ExportTabularDataReturn` so then we can write `return nil, err` here.

File: app/data_client.go
Comment (purplenicole730) [inline app/data_client.go:1251]: Make this return a pointer.

File: app/data_client.go
Comment (purplenicole730) [inline app/data_client.go:126]: Payload should be `map[string]interface{}`, not any.

**Review by purplenicole730 — CHANGES_REQUESTED** [review]

File: app/data_client_test.go
Comment (vijayvuyyuru) [reply app/data_client_test.go:402]: Ah ok cool, you answered my question. Thanks Nicole.

**Review by purplenicole730 — COMMENTED** [review]
This is looking good! Holding off on approving until we get the new `Any` type in here. It'll be in`common.go`. I'll notify when the `Any` type is available for you to use!

File: app/data_client.go
Comment (purplenicole730) [reply app/data_client.go]: If you merge main into this PR, you'll see a new `GenericProtoMessage` type in `common.go` that you can use!

Comment (vijayvuyyuru) [reply app/data_client.go]: Added!

**Review by purplenicole730 — APPROVED** [review]
LGTM

File: app/data_client.go
Comment (katiepeters) [inline app/data_client.go:135]: I can't claim to be an expert on writing Go SDKs, so this question might be more for my own understanding. How did we arrive at creating a `Next` method? Is this preferable over something like returning a channel?

File: app/data_client.go
Comment (katiepeters) [reply app/data_client.go:124]: I'm not sure I understand this. Could you explain the type choice?

File: app/data_client.go
Comment (vijayvuyyuru) [reply app/data_client.go:135]: I believe the idea is for the user to keep calling next to get readings as a stream. I based this on what we do for the robot logs streaming API and how it's implemented in the Go SDK.

File: app/data_client.go
Comment (purplenicole730) [reply app/data_client.go:113]: I'm assuming this was done since there are other data wrappers that call it `Return`. I personally would use the word `Response`, since it's a shadow type of the response. It's what we do for the Python SDK, as well as some other wrappers in the other app clients in the Go SDK.

File: app/data_client.go
Comment (purplenicole730) [reply app/data_client.go:135]: We're giving the user the option of what to do with the stream. Based on my conversation with others who are more familiar with Go standards, it's not common for Go to force users into a certain method of using something. The user can turn it into their own channel if that's what they prefer.

File: app/data_client.go
Comment (katiepeters) [reply app/data_client.go:124]: But that's not what we use in TabularDataByFilter?

File: app/data_client.go
Comment (katiepeters) [inline app/data_client.go:134]: Update use of `Return` in comments

File: app/data_client.go
Comment (purplenicole730) [reply app/data_client.go:124]: I haven't seen the actual proto version so this is the reference I made from Vijay's code, but the return types for `TabularDataByFilter` and `ExportTabularData` are different.

Comment (katiepeters) [reply app/data_client.go:124]: Ya, that's what I meant. `MethodParameters` inside `CaptureMetadata`

Comment (purplenicole730) [reply app/data_client.go:124]: @vijayvuyyuru Please hold onto this PR. I'll try to look into why we did it this way for `CaptureMetadata`. I didn't know that there was another case of `anypb.Any` being used here.

Comment (purplenicole730) [reply app/data_client.go:124]: Okay, @vijayvuyyuru sorry to give you more work, but I got some more background on this `Any` type, and realized that maybe going the same route that the `CaptureMetadata` did is better in terms of readability! I'll make those changes on my side as well. It shouldn't be too big of a fix for you, since `methodParamsFromProto` already exists in the data client. You should be able to directly call that method!

**Review by purplenicole730 — CHANGES_REQUESTED** [review]
Just putting this back so that I get a reminder to double check everything once you change the type for `MethodParameters` to `map[string]interface{}`

**Review by purplenicole730 — APPROVED** [review]
Sorry for making you go back and forth, but this looks good to me!

**Comment by vijayvuyyuru** [comment]
@purplenicole730 @katiepeters I updated this PR with the newest API changes. (ComponentName and ComponentType became ResourceName and ResourceSubtype) and (Additional parameters became a structpb, but still returning as a map of string to interfaces)

**Review by purplenicole730 — APPROVED** [review]
Looks good! Admittedly, I only took a quick glance to confirm the changes that you specified.

File: app/data_client.go
Comment (katiepeters) [inline app/data_client.go:1291]: Nit: rename `exportTabularDataResponseFromProto`

**Comment by katiepeters** [comment]
Looking pretty good now! Will approve once the method is actually available to use

**Review by katiepeters — APPROVED** [review]

**Pull request closed without merging**
